crates/sui-types/src/digests.rs

```diff
@@ -0,0 +1,566 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt;
+
+use crate::sui_serde::Readable;
+use fastcrypto::encoding::{Base58, Encoding};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, Bytes};
+
+/// A representation of a SHA3-256 Digest
+#[serde_as]
+#[derive(
+    Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, JsonSchema,
+)]
+pub struct Sha3Digest(
```

Comment:
did you consider something like

```
pub enum DigestType {
  Transaction,
  TransactionEffects,
  Object,
}

// preserve distinct types for each digest type
pub struct Sha3Digest<const Type:  DigestType> {
   ...
}

pub type TransactionDigest = Sha3Digest<DigestType::Transaction>;
// etc

impl<T> Sha3Digest<T> {
  // implementations shared
}

```

might save a lot of duplicated code and prevent inconsistent interfaces, if its not too much trouble.